lib/pages/community/community_store.dart

```diff
@@ -0,0 +1,114 @@
+import 'package:lemmy_api_client/v3.dart';
+import 'package:mobx/mobx.dart';
+
+import '../../util/async_store.dart';
+
+part 'community_store.g.dart';
+
+class CommunityStore = _CommunityStore with _$CommunityStore;
+
+abstract class _CommunityStore with Store {
+  final String instanceHost;
+  final String? communityName;
+  final int? id;
+
+  // ignore: unused_element
+  _CommunityStore.fromName({
+    required String this.communityName,
+    required this.instanceHost,
+  }) : id = null;
+  // ignore: unused_element
+  _CommunityStore.fromId({required this.id, required this.instanceHost})
+      : communityName = null;
+
+  final communityState = AsyncStore<FullCommunityView>();
+  final subscribingState = AsyncStore<CommunityView>();
+  final blockingState = AsyncStore<BlockedCommunity>();
+
+  @action
+  Future<void> fetch(Jwt? token) async {
+    await communityState.runLemmy(
+      instanceHost,
+      GetCommunity(
+        auth: token?.raw,
+        id: id,
+        name: communityName,
+      ),
+    );
+  }
+
+  @action
+  Future<void> refresh(Jwt? token) async {
+    await communityState.runLemmy(
+      instanceHost,
+      GetCommunity(
+        auth: token?.raw,
+        id: id,
+        name: communityName,
+      ),
+      refresh: true,
+    );
+  }
+
+  Future<void> block(Jwt token) async {
+    final communityView = communityState.asyncState.mapOrNull(
+      data: (value) => value.data.communityView,
+    );
+
+    if (communityView == null) {
+      throw StateError('FullCommunityView should be not null at this point');
+    }
+
+    final res = await blockingState.runLemmy(
+      instanceHost,
+      BlockCommunity(
+        communityId: communityView.community.id,
+        block: !communityView.blocked,
+        auth: token.raw,
+      ),
+    );
+
+    (communityState.asyncState as AsyncStateData).copyWith();
+
+    // communityState.asyncState.data.communityView = res.communityView
+    if (res != null) {
+      communityState.asyncState =
+          (communityState.asyncState as AsyncStateData<FullCommunityView>)
+              .copyWith(
+                  data: (communityState.asyncState
+                          as AsyncStateData<FullCommunityView>)
+                      .data
+                      .copyWith(communityView: res.communityView));
+    }
```

Comment:
```suggestion
     final state = communityState.asyncState;
    if (state is! AsyncStateData<FullCommunityView>) {
      throw StateError('communityState should be ready at this point');
    }

    final res = await blockingState.runLemmy(
      instanceHost,
      BlockCommunity(
        communityId: state.data.communityView.community.id,
        block: !state.data.communityView.blocked,
        auth: token.raw,
      ),
    );

    // communityState.asyncState.data.communityView = res.communityView
    if (res != null) {
      communityState.asyncState = state.copyWith(
        data: state.data.copyWith(communityView: res.communityView),
      );
    }
```

Will greatly improve ergonomics. Operate on types, not nullability

Reply:
In fact I would even consider adding a `.setData` methods (or of a similar name) to AsyncStore instead of directly modifying asyncState